File: project/core/module/rest-service/src/main/java/vdi/service/rest/conversion/DatasetCharacteristics.java
package vdi.service.rest.conversion;

import org.jetbrains.annotations.NotNull;
import vdi.service.rest.generated.model.DatasetCharacteristicsImpl;

public class DatasetCharacteristics extends DatasetCharacteristicsImpl {
  public DatasetCharacteristics(@NotNull vdi.model.meta.DatasetCharacteristics characteristics) {
    setStudyDesign(characteristics.getStudyDesign());
    setStudyType(characteristics.getStudyType());

    if (!characteristics.getCountries().isEmpty())
      setCountries(characteristics.getCountries());

    if (characteristics.getYears() != null)
      setYears(new SampleYearRange(characteristics.getYears()));

    if (!characteristics.getStudySpecies().isEmpty())
      setStudySpecies(characteristics.getStudySpecies());

    
  }
}

File: project/core/module/rest-service/src/main/kotlin/vdi/service/rest/server/conversion/DatasetCharacteristics.kt
package vdi.service.rest.server.conversion

import vdi.service.rest.generated.model.DatasetCharacteristicsImpl
import vdi.model.meta.DatasetCharacteristics as InternalType
import vdi.service.rest.generated.model.DatasetCharacteristics as RamlType

internal fun DatasetCharacteristics(props: InternalType): RamlType =
  DatasetCharacteristicsImpl().apply {
    studyDesign = props.studyDesign
    studyType = props.studyType

    props.countries
      .takeIf(List<*>::isNotEmpty)
      ?.also(::setCountries)

    props.years
      ?.let(::SampleYearRange)
      ?.also(::setYears)

    props.studySpecies
      .takeIf(List<*>::isNotEmpty)
      ?.also(::setStudySpecies)

    props.outcomes
      .takeIf(List<*>::isNotEmpty)
      ?.also(::setOutcomes)

    props.associatedFactors
      .takeIf(List<*>::isNotEmpty)
      ?.also(::setAssociatedFactors)

    props.participantAges
      ?.also(::setParticipantAges)

    props.sampleTypes
      .takeIf(List<*>::isNotEmpty)
      ?.also(::setSampleTypes)
  }